broker-core/src/main/java/io/zeebe/broker/subscription/message/state/WorkflowInstanceSubscriptionState.java

```diff
@@ -94,58 +99,108 @@ public void put(final WorkflowInstanceSubscription subscription) {
+    return offset;
+  }
+
+  private int writeSubscriptionKey(
+      MutableDirectBuffer buffer, int offset, final WorkflowInstanceSubscription subscription) {
+    final DirectBuffer messageName = subscription.getMessageName();
+    final int startOffset = offset;
+    final int expectedLength = getSubscriptionKeyLength(subscription);
+
     buffer.putLong(offset, subscription.getElementInstanceKey(), STATE_BYTE_ORDER);
     offset += Long.BYTES;
 
-    assert (2 * Long.BYTES) == offset
-        : "Offset problem: offset is not equal to expected key length";
+    offset = writeIntoBuffer(buffer, offset, messageName);
+    assert (offset - startOffset) == expectedLength : "End offset differs from expected length";
+
     return offset;
   }
 
-  private boolean readSubscription(long key, WorkflowInstanceSubscription subscription) {
-    final int readBytes = db.get(subscriptionColumnFamily, key, valueBuffer);
-    if (readBytes > 0) {
+  private void wrapSubscriptionKey(
+      DirectBuffer source, final WorkflowInstanceSubscription subscription) {
+    int offset = 0;
+
+    subscription.setElementInstanceKey(source.getLong(offset, STATE_BYTE_ORDER));
+    offset += Long.BYTES;
+
+    readIntoBuffer(source, offset, subscription.getMessageName());
+  }
+
+  private int getSubscriptionKeyLength(final WorkflowInstanceSubscription subscription) {
+    return Long.BYTES + Integer.BYTES + subscription.getMessageName().capacity();
+  }
+
+  private boolean readSubscription(WorkflowInstanceSubscription subscription) {
+    final int keyLength = writeSubscriptionKey(this.keyBuffer, 0, subscription);
+    final DirectBuffer keyBuffer = new UnsafeBuffer(this.keyBuffer, 0, keyLength);
+
+    final int readBytes = db.get(subscriptionColumnFamily, keyBuffer, valueBuffer);
+    if (readBytes != RocksDB.NOT_FOUND) {
       subscription.wrap(valueBuffer, 0, readBytes);
       return true;
-    } else {
-      return false;
     }
+
+    return false;
   }
 
-  public WorkflowInstanceSubscription getSubscription(long elementInstanceKey) {
-    final boolean found = readSubscription(elementInstanceKey, subscription);
+  public WorkflowInstanceSubscription getSubscription(
+      long elementInstanceKey, DirectBuffer messageName) {
+    subscription.setMessageName(messageName);
+    subscription.setElementInstanceKey(elementInstanceKey);
+    final boolean found = readSubscription(subscription);
+
     if (found) {
       return subscription;
     } else {
       return null;
     }
   }
 
+  public void visitElementSubscriptions(
+      long elementInstanceKey, WorkflowInstanceSubscriptionVisitor visitor) {
+    final UnsafeBuffer prefix = new UnsafeBuffer(keyBuffer, 0, Long.BYTES);
+    prefix.putLong(0, elementInstanceKey, STATE_BYTE_ORDER);
+
+    db.forEachPrefixed(
+        subscriptionColumnFamily,
+        prefix,
+        ((entry, control) -> {
+          wrapSubscriptionKey(entry.getKey(), subscription);
+          if (readSubscription(subscription)) {
```

Comment:
We should throw an exception if the subscription is not read. Otherwise, we may oversee an inconsistent state.